File: go/exercises/practice/unit_test_9770fc8a_fd13_44f6_9008_b6a130faf3e2/txn_manager/.meta/dtm.go
package txn_manager

import (
	"errors"
	"sync"
	"time"
)

// Participant represents a participant in the distributed transaction.
type Participant struct {
	ID       string
	Prepare  func() error
	Commit   func() error
	Rollback func() error
}

// TransactionState represents the state of a transaction.
type TransactionState int

const (
	Unknown TransactionState = iota
	Committed
	RolledBack
	InProgress
)

// LogEntry represents a persistent log entry for transaction state changes.
type LogEntry struct {
	TXID  string
	State TransactionState
}

// DTM represents the distributed transaction manager.
type DTM struct {
	transactions   map[string]TransactionState
	logs           []LogEntry
	mutex          sync.Mutex
	prepareTimeout time.Duration
}

// NewDTM creates a new instance of DTM.
func NewDTM() *DTM {
	return &DTM{
		transactions:   make(map[string]TransactionState),
		logs:           []LogEntry{},
		prepareTimeout: 1 * time.Second,
	}
}

// NewDTMFromLog creates a new DTM instance from a persisted log.
func NewDTMFromLog(logs []LogEntry) *DTM {
	dtm := &DTM{
		transactions:   make(map[string]TransactionState),
		logs:           logs,
		prepareTimeout: 1 * time.Second,
	}
	for _, entry := range logs {
		dtm.transactions[entry.TXID] = entry.State
	}
	return dtm
}

// ExportLog returns the current persistent log.
func (d *DTM) ExportLog() []LogEntry {
	d.mutex.Lock()
	defer d.mutex.Unlock()
	copiedLog := make([]LogEntry, len(d.logs))
	copy(copiedLog, d.logs)
	return copiedLog
}

// SetPrepareTimeout sets the timeout for the preparation phase for each participant.
func (d *DTM) SetPrepareTimeout(timeout time.Duration) {
	d.mutex.Lock()
	defer d.mutex.Unlock()
	d.prepareTimeout = timeout
}

// ProcessTransaction coordinates the 2PC transaction process.
func (d *DTM) ProcessTransaction(txid string, participants []Participant) error {
	d.mutex.Lock()
	// Idempotency: if transaction already committed or rolled back, return existing result.
	if state, exists := d.transactions[txid]; exists && (state == Committed || state == RolledBack) {
		d.mutex.Unlock()
		return nil
	}
	// Mark transaction as in progress.
	d.transactions[txid] = InProgress
	d.logs = append(d.logs, LogEntry{TXID: txid, State: InProgress})
	d.mutex.Unlock()

	// Phase 1: Prepare Phase.
	prepareResults := make(chan error, len(participants))
	var wg sync.WaitGroup

	for _, p := range participants {
		wg.Add(1)
		go func(part Participant) {
			defer wg.Done()
			ch := make(chan error, 1)
			go func() {
				err := part.Prepare()
				ch <- err
			}()
			select {
			case err := <-ch:
				prepareResults <- err
			case <-time.After(d.prepareTimeout):
				prepareResults <- errors.New("prepare timeout")
			}
		}(p)
	}
	wg.Wait()
	close(prepareResults)

	prepareFailed := false
	for err := range prepareResults {
		if err != nil {
			prepareFailed = true
			break
		}
	}

	if prepareFailed {
		d.rollbackTransaction(txid, participants)
		return errors.New("prepare phase failed, transaction rolled back")
	}

	// Phase 2: Commit Phase.
	for _, p := range participants {
		if err := p.Commit(); err != nil {
			d.rollbackTransaction(txid, participants)
			return errors.New("commit phase failed, transaction rolled back")
		}
	}

	d.mutex.Lock()
	d.transactions[txid] = Committed
	d.logs = append(d.logs, LogEntry{TXID: txid, State: Committed})
	d.mutex.Unlock()
	return nil
}

func (d *DTM) rollbackTransaction(txid string, participants []Participant) {
	for _, p := range participants {
		_ = p.Rollback()
	}
	d.mutex.Lock()
	d.transactions[txid] = RolledBack
	d.logs = append(d.logs, LogEntry{TXID: txid, State: RolledBack})
	d.mutex.Unlock()
}

// GetTransactionState returns the current state of the transaction.
func (d *DTM) GetTransactionState(txid string) TransactionState {
	d.mutex.Lock()
	defer d.mutex.Unlock()
	if state, exists := d.transactions[txid]; exists {
		return state
	}
	return Unknown
}

// SimulateCrash clears the in-memory transaction states to simulate a crash.
func (d *DTM) SimulateCrash() {
	d.mutex.Lock()
	defer d.mutex.Unlock()
	d.transactions = make(map[string]TransactionState)
}

// Recover replays the log to finalize any in-flight transactions.
func (d *DTM) Recover() error {
	d.mutex.Lock()
	defer d.mutex.Unlock()
	for _, entry := range d.logs {
		if entry.State == InProgress {
			// For simplicity, roll back any in-progress transaction.
			d.transactions[entry.TXID] = RolledBack
			d.logs = append(d.logs, LogEntry{TXID: entry.TXID, State: RolledBack})
		} else {
			d.transactions[entry.TXID] = entry.State
		}
	}
	return nil
}

File: python/exercises/practice/unit_test_3ea01c01_05cb_47a0_9465_625fae9f34e0/dtm_consistency/.meta/dtm.py
import threading
import os
from enum import Enum

class TransactionState(Enum):
    ACTIVE = "ACTIVE"
    PREPARED = "PREPARED"
    COMMITTED = "COMMITTED"
    ROLLED_BACK = "ROLLED_BACK"
    NON_EXISTENT = "NON_EXISTENT"

class DTM:
    def __init__(self, log_file_path):
        self.log_file_path = log_file_path
        self.transactions = {}
        self.lock = threading.Lock()
        self._initialize_log_file()

    def _initialize_log_file(self):
        with open(self.log_file_path, 'a'):
            pass

    def _log_action(self, action, transaction_id, service_id=None):
        with open(self.log_file_path, 'a') as log_file:
            if service_id is not None:
                log_file.write(f"{action} {transaction_id} {service_id}\n")
            else:
                log_file.write(f"{action} {transaction_id}\n")

    def begin_transaction(self, transaction_id):
        with self.lock:
            if transaction_id in self.transactions:
                raise ValueError("Transaction ID already exists")
            self.transactions[transaction_id] = {
                'state': TransactionState.ACTIVE,
                'prepared_services': set()
            }
            self._log_action("BEGIN", transaction_id)

    def prepare(self, transaction_id, service_id):
        with self.lock:
            if transaction_id not in self.transactions:
                raise ValueError("Transaction does not exist")
            if self.transactions[transaction_id]['state'] != TransactionState.ACTIVE:
                raise ValueError("Transaction is not in ACTIVE state")
            
            self.transactions[transaction_id]['prepared_services'].add(service_id)
            self._log_action("PREPARE", transaction_id, service_id)
            
            if len(self.transactions[transaction_id]['prepared_services']) > 0:
                self.transactions[transaction_id]['state'] = TransactionState.PREPARED

    def commit(self, transaction_id):
        with self.lock:
            if transaction_id not in self.transactions:
                raise ValueError("Transaction does not exist")
            if self.transactions[transaction_id]['state'] != TransactionState.PREPARED:
                raise ValueError("Transaction is not in PREPARED state")
            
            self.transactions[transaction_id]['state'] = TransactionState.COMMITTED
            self._log_action("COMMIT", transaction_id)

    def rollback(self, transaction_id):
        with self.lock:
            if transaction_id not in self.transactions:
                raise ValueError("Transaction does not exist")
            if self.transactions[transaction_id]['state'] not in [TransactionState.ACTIVE, TransactionState.PREPARED]:
                raise ValueError("Transaction cannot be rolled back in current state")
            
            self.transactions[transaction_id]['state'] = TransactionState.ROLLED_BACK
            self.transactions[transaction_id]['prepared_services'].clear()
            self._log_action("ROLLBACK", transaction_id)

    def get_transaction_state(self, transaction_id):
        with self.lock:
            if transaction_id not in self.transactions:
                return TransactionState.NON_EXISTENT.value
            return self.transactions[transaction_id]['state'].value